webapi/src/main/scala/org/knora/webapi/messages/admin/responder/KnoraResponseADM.scala

```diff
@@ -10,4 +10,6 @@ import org.knora.webapi.messages.traits.Jsonable
 /**
  * A trait for Knora Admin response messages. Any response message can be converted into JSON.
  */
-trait KnoraResponseADM extends Jsonable
+trait KnoraResponseADM extends Jsonable {
+  def format: KnoraResponseADM
```

Comment:
AFAIU the `KnoraResponseADM` is supposed represent a  data structure for a response ready to be returned to the http client, the `Jsonable` trait handles the conversion to the json string, am I right?

You are calling the `format` method in a single the place right before the response is rendered as a `HttpResponse` out to the client.  Most of the times the `format` does return the unchanged instance in some cases the data has to be changed. In these cases the implementation of this trait  not only contains the data but also the knowledge and behaviour of transforming itself to different data. Sometimes this also depends on the infamous `StringFormatter. 

IMHO for me this is confusing. The `KnoraResponseADM` is not a data structure anymore. Any instance of it cannot simply be rendered out as json. Turning an instance into a json string must be done with `someKnoraResponseADMInstance.format.toJsValue.asJsObject.compactPrint`.

Before your change rendering the json (`.toJsValue.asJsObject.compactPrint`) was already complicated, now the `.format` in the beginning can  be easily forgotten. Also, the method name does not reveal what `format` does. Even if it would come with documentation of what it does and/or under which circumstances this method must be called it is still complicated to use and understand and therefore error prone.

Also data structures should not depend on the `StringFormatter` I think.

I do not believe that documenting this behaviour and the dependencies that come along with it would solve the issues I have with this approach. Wouldn't it make more sense to assemble the data structure with the correct data from the beginning? Or maybe have a dedicated class which would encapsulate the transformation?

Reply:
Thanks for the feedback. I can understand your concerns  and I am happy to address them - but I'm not sure how to solve it (see my questions below). Apart from that, it is handled in the same way in V2 - although this doesn't mean that we need to do it the same way in ADM, of course. The same pattern also applies for the `toJsValue` as far as I understand. So the response data structures know themselves how they need to be transformed into JSON. Is this so different from knowing how to be transformed into the external format?

>>Wouldn't it make more sense to assemble the data structure with the correct data from the beginning? 
If I understand you correctly, this would mean that the responder would need to format each (affected) response before returning it. I think this would be far more error prone because it can easily be forgotten. Instead, calling format in the `RouteUtilADM` will handle all responses in the same way in one location.

>>Or maybe have a dedicated class which would encapsulate the transformation?
I imagine such a service (?) would pattern match on the response type and format it accordingly? But if something changes in the response (i.e. the data structure) than it could be easily forgotten to adjust it in this service. This doesn't seem robust enough to me. Also, this service would be called in the same location as `format` is now called, right? Just before the response is turned into a `HttpResponse`. So, I don't really see the whole benefit of it, yet.

Reply:
> So the response data structures know themselves how they need to be transformed into JSON. Is this so different from knowing how to be transformed into the external format?

Yes, taking a data structure and rendering it out as json/yaml/xml is simply serialising the exact same data. The "external format" is changing the data itself.  IMHO, internal and external `IRI`s should not even be allowed to be the same type in our domain logic because the represent two different things. If I would give you an instance of `IRI` aka. `String` you would not know where it may be used, is it internal/external? The compiler does not help us, the type system does not help us, only `StringFormatter` could check it for you during runtime. If I give you an instance of the the `KnoraResponseADM` you would not know if it was already "formatted" or not, does it contain external or internal `IRI`s?

> Instead, calling format in the RouteUtilADM will handle all responses in the same way in one location.
This is an advantage only if you assume that the `RouteUtilADM` will always be the same and only place where we would want to use the response objects. Hence I think we should not consider such "optimisations" in the current implementation for now - especially because we will use these response models  in the zio-http routes as well.

>  If I understand you correctly, this would mean that the responder would need to format each (affected) response before returning it. I think this would be far more error prone because it can easily be forgotten. 
It is not "formatting" the data - the `format` method is replacing a value with another value in the same place. The different values are used for different purposes. I think confusion arises from the fact that the `KnoraResponseADM` serves these two different purposes:
1. it is a response from a responder and may be used internally by other responders
2. it is a response to a http request from the outside world